crossbar/worker/proxy.py

```diff
@@ -60,6 +57,284 @@ class ProxyWorkerProcess(worker.NativeWorkerProcess):
+                self,
+                self._realm,
+                self._authid,
+                self._authrole,
+            )
+            backend_d.addCallback(_backend_connected)
+            backend_d.addErrback(_backend_failed)
+        else:
+            # 'msg' is a Deny or a Challenge, so communicate that to
+            # the client before we try connecting to the backend
+            self.transport.send(msg)
+
+    def _do_authentication(self, realm, details):
+        if not realm:
+            return types.Deny(
+                ApplicationError.NO_SUCH_REALM,
+                message='no realm requested',
+            )
+        if not self._controller.has_realm(realm):
+            return types.Deny(
+                ApplicationError.NO_SUCH_REALM,
+                message='no realm "{}" exists on this router'.format(realm),
+            )
+        if not self._controller.can_map_backend(details.pending_session, realm, details.authid, details.authrole):
+            # XXX maybe want better errors here? maybe can_map should return a Deny?
+            return types.Deny(
+                "wamp.error.runtime_error",
+                message='Proxy cannot map session to a backend (realm="{}" role="{}")'.format(
+                    realm,
+                    details.authrole,
+                )
+            )
+
+        if details.authmethods and 'anonymous' not in details.authmethods:
+            return types.Deny(
+                ApplicationError.NO_AUTH_METHOD,
+                message='cannot authenticate using any of the offered authmethods {}'.format(details.authmethods),
+            )
+
+        from crossbar.router.auth import AUTHMETHOD_MAP
+        PendingAuthKlass = AUTHMETHOD_MAP['anonymous']
+        self._pending_auth = PendingAuthKlass(
+            details.pending_session,
+            {"type": "proxy"},  # what else goes in _transport_info
+            self._controller,
+            {
+                'type': 'static',
+                'role': details.authrole,
+                'authid': util.generate_serial_number(),
+            }
+        )
+        self._authrole = details.authrole
+        self._authprovider = 'anonymous'
+        self._authextra = details.authextra
+
+        return self._pending_auth.hello(realm, details)
+
+    def forward_message(self, msg):
+        self.transport.send(msg)
+
+    def onClose(self, wasClean):
+        """
+        Callback fired when the transport has been closed.
+
+        :param wasClean: Indicates if the transport has been closed regularly.
+        :type wasClean: bool
+        """
+        # print("ProxySession.onClose: {}".format(wasClean))
+        if self._backend_session is not None:
+            try:
+                self._backend_session.leave()
+            except Error:
+                pass
+            self._backend_session = None
+            self.transport = None
+
+
+ITransportHandler.register(ProxySession)
+
+
 class BackendProxySession(Session):
```

Comment:
This looks already great! as far as I see, the only piece missing is `onChallenge` and adding WAMP-cryptosign authentication, using the node private key and forwarding the incoming (already authenticated) principal `authid`, `authrole` and `realm` in the `authextra` used in the backend connection. All the other pieces are already there. Awesome=)

Reply:
I just realize: for now, it's even simpler than that: because if the backend router worker listens on UDS, the filesystem is authenticating the incoming proxy connections, and all the backend authenticator needs to do is pick out authid, authrole and realm from authextra and return that as principal.

Reply:
Yeah. And I don't think it's worth supporting anything besides "nothing, it's UDS" or "cryptosign-proxy" for the actual backend connections, right? (I mean, `wampcra` and `wamp-scram` or tickets)

Reply:
>  And I don't think it's worth supporting anything besides "nothing, it's UDS" or "cryptosign-proxy" for the actual backend connections, right?

yes, agreed, only 2 methods: nothing and cryptosign